### DISCUSSION TITLE
GitHub should support GitHub OIDC

### DISCUSSION THREAD

**Discussion by captn3m0** [community/community#27300] [Other Feature Feedback, Questions, & Ideas]
As it stands, GitHub OIDC is only supported by a few third-parties as a means of authorization: AWS, Azure, GCP, Vault.

A few more providers are adding support, such as PyPi.

However, GitHub owned properties should also support GitHub OIDC based permissions. This would primarily be:

1. GitHub Apps
2. ~~NPM~~

### GitHub Apps

To move to a modern zero-trust keyless environment, the default GitHub Action works great, but it cannot be used to access other repositories. GitHub Apps provide the most secure way of granting access to GitHub itself. However, this is bound by a Client Secret, which needs to be stored in every application as a secret separately.

Having a simple policy (such as in Rego) attached to a GitHub app would provide a way for a GitHub Token to escalate its permissions to a GitHub App, without any secrets being stored. This would provide a way for GitHub Actions to authenticate to GitHub using GitHub OIDC

### ~~NPM~~

By supporting GitHub OIDC for publishing packages, GitHub can finally make keyless publication a reality for NPM. The vast majority of packages on GitHub are already published via GitHub, and having a way to sign and publish packages to NPM easily will be a great win.

**Update**: NPM shipped Trusted Publishing in Jul 2025: https://github.blog/changelog/2025-07-31-npm-trusted-publishing-with-oidc-is-generally-available/

**Comment by richardfan1126** [comment]
This also allows a repository workflow accessing another repository without using any secret

This is a very common use case (e.g. a meta IaC repo managing the configs of all other repos) and is supported in GitLab already, by [job token allowlist](https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html#add-a-group-or-project-to-the-job-token-allowlist)

**Comment by Freyert** [comment]
GitOps applications would really benefit from the addition of OIDC to GitHub apps.

GitHub apps would then provide a good experience for the "machine user" use case that is really missing from GitHub, but found in so many other places.
